Add groups to win/lose criteria

tobylane opened the pull request:
**Describe what the proposed change does**
- Collect level goals into their groups
- Check each group for win and lose game endings
- Make the progress report dialog show the most relevant group

Fixes #232

TheCycoONE commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -28,7 +28,7 @@ function UIProgressReport:UIProgressReport(ui)
 
   local world = self.ui.app.world
-  local hospital = ui.hospital
+  self.world  = world
```
not much reason for the local world above then right?

lewri commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,323 @@
+class "WinCriteria"
+--@type WinCriteria
+
+--!param hospital (Hospital) The hospital of the tests.
+--!param table (table) the win_criteria table for one group and one criteria
+--!return If report is false, return the gap from hospital value to win value, and the win value.
+-- if true, return the gap from the hospital value to boundary, and the boundary.
```
On doc, this only will apply to the top function. I fully understand why you made this decision, but for our documentation system it leaves gaps for sure.

Alberth289346 replied:
If you merge the function computations into one as suggested elsewhere, this doc can be attached to that function.

lewri commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,323 @@
+  if (rep - target) * max_min >= 0 then
+    return math.abs(rep - target), target
+  end
+end
+
+function WinCriteria.balance(hospital, table, report)
+  local target = report and table.boundary or table.win_value
+  local bal = hospital.balance - hospital.loan
+  local max_min = table.max_min_win == 1 and 1 or -1
+  if (bal - target) * max_min >= 0 then
+    return math.abs(bal - target), target
+  end
+end
+
+function WinCriteria.percentage_cured(hospital, table, report)
+  local target = report and table.boundary or table.win_value
+  local percentage_cured = hospital.percentage_cured
+  local max_min = table.max_min_win == 1 and 1 or -1
+  if (percentage_cured - target) * max_min >= 0 then
+    return math.abs(percentage_cured - target), target
+  end
+end
+
+function WinCriteria.num_cured(hospital, table, report)
+  local target = report and table.boundary or table.win_value
+  local num_cured = hospital.num_cured
+  local max_min = table.max_min_win == 1 and 1 or -1
+  if (num_cured - target) * max_min >= 0 then
+    return math.abs(num_cured - target), target
+  end
+end
+
+function WinCriteria.value(hospital, table, report)
+  local target = report and table.boundary or table.win_value
+  local value = hospital.value
+  local max_min = table.max_min_win == 1 and 1 or -1
+  if (value - target) * max_min >= 0 then
+    return math.abs(value - target), target
+  end
+end
```
You're repeating a lot of the same execution here from a code perspective. If there is a common thread of actions that could be condensed into one local or private function then I'd recommend that. Then call that function each time.

lewri commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -244,5 +227,9 @@ function UIProgressReport:afterLoad(old, new)
     self.panel_sprites = gfx:loadSpriteTable("QData", "Rep02V", true, palette)
   end
 
+  if old < 182 then
```
Needs a bump

lewri commented on CorsixTH/Lua/world.lua:
```diff
@@ -2801,6 +2659,12 @@ function World:afterLoad(old, new)
     self:initWallTypes()
   end
 
+  if old < 182 then
```
Need a bump

lewri commented on CorsixTH/Lua/world.lua:
```diff
@@ -299,69 +288,11 @@ function World:toggleInformation()
-        world_goals[world_goals[crit_name].number].group = values.Group
-      end
-    end
+  if not (self.app.using_demo_files or self.free_build_mode) then
```
The demo definitely has end conditions does it not?
Edit: Looks like CorsixTH actually loads the base_config, which give no win/lose condition so this isn't wrong.
But a different issue for later is that the Demo has a defined endCondition in the readme:
```txt
You have to cure 20 patients in a year in this demo so look after your
patients well and provide for their every need.
```
And in the demo itself from DOSBOX
![image](https://github.com/CorsixTH/CorsixTH/assets/20030128/f80942db-e112-47c9-81c5-eca6866e9ac7)

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,323 @@
+  end
+end
+
+function WinCriteria.value(hospital, table, report)
```
You may want to rename this one to `hospitalValue`, since `value` is quite generic.

tobylane replied:
It uses the hospital attribute name. I'll add a comment to clarify that.

tobylane commented:
Converted the identical set of tests into a generic test, with the ability to add specific tests later.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,285 @@
+--!param table (table) the win_criteria table for one group and one criteria
+--!return If report is false, return the gap from hospital value to win value, and the win value.
+-- if true, return the gap from the hospital value to boundary, and the boundary.
+function WinCriteria.generic(hospital, crit_name, table, report)
```
Does this work? I am missing the `local WinCriteria = _G["WinCriteria"]` line that makes the `WinCriteria` used here available.

tobylane replied:
It does, even the EndCondition one is optional. I've added them.

lewri commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,285 @@
+-- Value is the hospital value.
+--!return If report is false, return the gap from hospital value to lose value, and the lose value
+-- if true, return the gap from the hospital value to boundary, and the boundary.
+function LoseCriteria.generic(hospital, attribute, table, report)
```
Generic is a worrisome function name. Why not `LoseCriteria:testCriterion`?

tobylane commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,302 @@
+  local target = report and table.boundary or table.lose_value
+  local measure = hospital[attribute]
+  local max_min = table.max_min_lose == 1 and 1 or -1
+  if (measure - target) * max_min > 0 then
```
This isn't the right comparison, it doesn't warn the user early enough in cases like #232.

tobylane commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,244 @@
+--!param hospital (Hospital) The hospital of the tests.
+--!return best (number) The number of the best group.
+function EndConditions:findBestWinGroup(hospital)
+assert(self.highest_group, "error")
```
Remove this.

tobylane reviewed (commented):
The generated parts of EndConditions - win_goals, lose_goals and highest_group, aren't kept in savegames, what's needed?

Added a get function for when a criterion isn't an attribute of the hospital, and a make_getter for when it is. This will make it possible to add more variety of new criteria later. Why does make_getter need to be persistable?

tobylane commented:
Fixed the issues in my previous comment, and added `EndConditions:getAttribute` to make it easier to change the get attribute functions later.

Fixes #232, #2432

tobylane commented:
This puts the persisted data in world, a better fix.

Rebased, ready.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,254 @@
+--[[ Copyright (c) 2023 Toby "tobylane"
```
Wrong year?

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,254 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE. --]]
+
+-- The order of this table relates to the "Winning and Losing Conditions" section
```
```suggestion
-- The order of this table is the same as the goals in the "Winning and Losing Conditions" section
```
I am guessing this is what you're aiming for.

"relates to" is too free, you should give enough information such that anyone can check that the table below is accurate and/or fix it if the table is incorrect for whatever reason.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,254 @@
+      end
+    end
+  end
+  return self, win_goals, lose_goals, highest_group
```
The class has no data (no `self.x` values). As such there is no need to instantiate the class.

Code will work fine if you treat it as a bunch of static functions.

As an example, consider `function EndConditions:checkWinGroup(hospital, win_table, findbest)`. Change it to `function EndConditions.checkWinGroup(hospital, win_table, findbest)` (colon becomes a dot).

As `self` doesn't exist without instantiating the class, you change calls to the function from eg `local reason = self:checkWinGroup(hospital, tbl)` to `local reason = EndConditions.checkWinGroup(hospital, tbl)` ( `self:` becomes `EndConditions.`). 

What happens is close to
``` lua
EndConditions = {}
EndConditions.checkWinGroup = function (hospital, win_table, findbest) ... end

local reason = EndConditions.checkWinGroup(hospital, tbl)
```
That is, just a bunch of functions stored in a table. Call such a function is pointing at it, and then append `(....)` to say "execute it", as usual with a function.

Alberth289346 replied:
The other direction of course is to have an instance of this class, and then you can store the data that you now return from the constructor. It then becomes available for all other methods so you don't need to drag that around in other variables.

tobylane replied:
I did that at first, but it wasn't saved. Is it better to have it save in World, or find out why it can't save in EndConditions?

TheCycoONE replied:
Look into why the class isn't persisting.

Alberth289346 replied:
Indeed, I see no reason why you couldn't store data in a class instance in world.

Alberth289346 commented on CorsixTH/Lua/world.lua:
```diff
@@ -299,69 +288,11 @@ function World:toggleInformation()
-
-  -- Order the criteria (some icons in the progress report shouldn't be next to each other)
-  table.sort(world_goals, function(a,b) return a.criterion < b.criterion end)
-  self.goals = world_goals
-  self.winning_goal_count = winning_goal_count
+  if self.free_build_mode then return end
+  self.endconditions, self.win_goals, self.lose_goals, self.highest_group =
+      EndConditions:EndConditions(self.map.level_config, self.map.level_number)
```
If doing the checks in EndConditions is the goal, why are you not storing the data needed for it inside that instance?

If you want to do it like you propose, at least wrap a table around all the data you put in world, so there is only one variable in the class containing the data.

We already have too many variables in world, please avoid making (many) more.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,254 @@
+  end
+
+  if findbest then
+    if met_count > 0 and total_count > 0 then
```
`total_count` is always at least `met_count`, so if the former holds, there is no need to test the latter.

You can add an assert if you are afraid this relation may get broken.

Alberth289346 commented on CorsixTH/Lua/endconditions.lua:
```diff
@@ -0,0 +1,254 @@
+      end
+    end
+  end
+  local lose = level_config.lose_criteria
+  if lose then
+    for _, values in pairs(lose) do
+      if values.Criteria ~= 0 then
+        local crit_name = local_criteria_variable[values.Criteria].name
+        if not lose_goals[values.Group] then lose_goals[values.Group] = {} end
+        lose_goals[values.Group][crit_name] = {
+          name = crit_name,
+          lose_value = values.Value,
+          boundary = values.Bound,
+          criterion = values.Criteria,
+          max_min_lose = values.MaxMin,
+          icon = local_criteria_variable[values.Criteria].icon,
+          formats = local_criteria_variable[values.Criteria].formats,
+          start = start[crit_name] or 0,
+        }
+        if values.Group > highest_group then highest_group = values.Group end
+      end
+    end
+  end
```
Looks like a copy of the win-criteria code just above it, except for the outer names `lose`, `lose_criteria`, and `lose_goals`.

If so, refactor this function to a private function and pass in the variables I just listed to the function, so you avoid copy/pasting the entire body.

Alberth289346 commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -49,52 +48,35 @@ function UIProgressReport:UIProgressReport(ui)
       -- string.format("%.0f", x) is not suitable due to %d (num) param in _S string
       local tooltip
-      if world.level_criteria[tab.criterion].formats == 2 then
+      if crit_table.formats == 2 then
         tooltip = _S.tooltip.status[crit_name]:format(math.floor(res_value), math.floor(cur_value))
       else
         tooltip = _S.tooltip.status[crit_name]:format(math.floor(res_value))
       end
```
Wouldn't `formats` be better named as `number_of_values` ?

tobylane replied:
I have plans for other formats that are two values,
```lua
       elseif crit_table.formats == 3 then
         tooltip = tooltip:format(math.floor(res_value) / 100, math.floor(cur_value) / 100)
```

Alberth289346 commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -49,52 +48,35 @@ function UIProgressReport:UIProgressReport(ui)
-        end
-      elseif not world_goals[crit_name].win_value then
-        world_goals[crit_name].visible = false
+  if not world.free_build_mode then
```
Can't you have empty win and lose goals instead?

That avoids having this check here.

tobylane commented:
Addressed feedback, improved comments and rebased.

Alberth289346 commented on CorsixTH/Lua/dialogs/fullscreen/progress_report.lua:
```diff
@@ -49,53 +48,34 @@ function UIProgressReport:UIProgressReport(ui)
-    if crit_name == "balance" then
-      cur_value = cur_value - hospital.loan
+  local crit_data = world.endconditions:generateReportTable(hospital)
+  for i, crit_table in ipairs(crit_data) do
```
`i` seems unused, by convention we write that as `_`.

Alberth289346 commented:
LGTM
